### PR TITLE
Aws asff

### PR DESCRIPTION
**Pull request by brianterry**
This PR adds support for AWS Security hub.

It has the following features:

- Sends findings in batches of 100
- Currently sends findings with the state of **Warning** and **Failed**

### CODE REVIEW

File: internal/findings/publisher.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		client: client,
+	}
+}
+
+// PublishFinding publishes findings to AWS Security Hub Service
+func (p *Publisher) PublishFinding(finding []*securityhub.AwsSecurityFinding) (*PublisherOutput, error) {
+	o := PublisherOutput{}
+	i := securityhub.BatchImportFindingsInput{}
+	i.Findings = finding
+	var errs error = nil
```
Comment (lizrice) [inline internal/findings/publisher.go:42]:
```suggestion
	var errs error
```
It will be initialized to nil by default

File: check/controls.go [hunk 2]
```diff
@@ -185,6 +200,69 @@ func (controls *Controls) JUnit() ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+// AASF encodes the results of last run to AWS Security Finding Format(AASF).
```
Comment (lizrice) [inline check/controls.go:203]:
```suggestion
// ASFF encodes the results of last run to AWS Security Finding Format(ASFF).
```

Comment (lizrice) [reply check/controls.go:203]: Prestty sure it should be ASFF, right? (https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-findings-format.html)

File: check/controls.go [hunk 3]
```diff
@@ -185,6 +200,69 @@ func (controls *Controls) JUnit() ([]byte, error) {
...
+func (controls *Controls) AASF() ([]*securityhub.AwsSecurityFinding, error) {
```
Comment (lizrice) [inline check/controls.go:204]:
```suggestion
func (controls *Controls) ASFF() ([]*securityhub.AwsSecurityFinding, error) {
```

Comment (lizrice) [reply check/controls.go:204]: this typo in the method name would need corecting elsewhere too, I am not sure I will catch them all

File: check/controls_test.go [hunk 4]
```diff
@@ -357,3 +362,103 @@ func assertEqualGroupSummary(t *testing.T, pass, fail, info, warn int, actual *G
 	assert.Equal(t, info, actual.Info)
 	assert.Equal(t, warn, actual.Warn)
 }
+
+func TestControls_AASF(t *testing.T) {
```
Comment (lizrice) [inline check/controls_test.go:366]:
```suggestion
func TestControls_ASFF(t *testing.T) {
```

File: check/controls_test.go [hunk 5]
```diff
@@ -357,3 +362,103 @@ func assertEqualGroupSummary(t *testing.T, pass, fail, info, warn int, actual *G
...
+			viper.Set("AWS_ACCOUNT", "foo account")
+			viper.Set("CLUSTER_ARN", "foo Cluster")
+			controls := &Controls{
+				ID:      tt.fields.ID,
+				Version: tt.fields.Version,
+				Text:    tt.fields.Text,
+				Groups:  tt.fields.Groups,
+				Summary: tt.fields.Summary,
+			}
+			got, _ := controls.AASF()
```
Comment (lizrice) [inline check/controls_test.go:455]:
```suggestion
			got, _ := controls.ASFF()
```

File: check/controls_test.go [hunk 6]
```diff
@@ -357,3 +362,103 @@ func assertEqualGroupSummary(t *testing.T, pass, fail, info, warn int, actual *G
...
+			tt.want[0].CreatedAt = got[0].CreatedAt
+			tt.want[0].UpdatedAt = got[0].UpdatedAt
+			tt.want[0].Id = got[0].Id
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Controls.AASF() = %v, want %v", got, tt.want)
```
Comment (lizrice) [inline check/controls_test.go:460]:
```suggestion
				t.Errorf("Controls.ASFF() = %v, want %v", got, tt.want)
```

File: cmd/common.go [hunk 7]
```diff
@@ -373,6 +373,10 @@ func writeOutput(controlsCollection []*check.Controls) {
 		writePgsqlOutput(controlsCollection)
 		return
 	}
+	if aASF {
+		writeAASFOutput(controlsCollection)
```
Comment (lizrice) [inline cmd/common.go:377]:
```suggestion
		writeASFFOutput(controlsCollection)
```

File: cmd/common.go [hunk 8]
```diff
@@ -404,6 +408,18 @@ func writePgsqlOutput(controlsCollection []*check.Controls) {
 	}
 }
 
+func writeAASFOutput(controlsCollection []*check.Controls) {
```
Comment (lizrice) [inline cmd/common.go:411]:
```suggestion
func writeASFFOutput(controlsCollection []*check.Controls) {
```

File: cmd/common.go [hunk 9]
```diff
@@ -404,6 +408,18 @@ func writePgsqlOutput(controlsCollection []*check.Controls) {
...
+	for _, controls := range controlsCollection {
+		out, err := controls.AASF()
```
Comment (lizrice) [inline cmd/common.go:413]:
```suggestion
		out, err := controls.ASFF()
```

File: cmd/common.go [hunk 10]
```diff
@@ -404,6 +408,18 @@ func writePgsqlOutput(controlsCollection []*check.Controls) {
...
+		if err != nil {
+			exitWithError(fmt.Errorf("failed to output to AASF: %v", err))
+		}
+		if err := writeFinding(out); err != nil {
+			exitWithError(fmt.Errorf("failed to output to AASF: %v", err))
```
Comment (lizrice) [inline cmd/common.go:418]:
```suggestion
			exitWithError(fmt.Errorf("failed to output to ASFF: %v", err))
```

File: cmd/common.go [hunk 11]
```diff
@@ -373,6 +373,10 @@ func writeOutput(controlsCollection []*check.Controls) {
 		writePgsqlOutput(controlsCollection)
 		return
 	}
+	if aASF {
```
Comment (lizrice) [inline cmd/common.go:376]:
```suggestion
	if aSFF {
```

File: cmd/root.go [hunk 12]
```diff
@@ -42,6 +42,7 @@ var (
 	jsonFmt             bool
 	junitFmt            bool
 	pgSQL               bool
+	aASF                bool
```
Comment (lizrice) [inline cmd/root.go:45]:
```suggestion
	aSFF                bool
```

File: cmd/root.go [hunk 13]
```diff
@@ -156,6 +157,7 @@ func init() {
 	RootCmd.PersistentFlags().BoolVar(&jsonFmt, "json", false, "Prints the results as JSON")
 	RootCmd.PersistentFlags().BoolVar(&junitFmt, "junit", false, "Prints the results as JUnit")
 	RootCmd.PersistentFlags().BoolVar(&pgSQL, "pgsql", false, "Save the results to PostgreSQL")
+	RootCmd.PersistentFlags().BoolVar(&aASF, "assf", false, "Send the results to AWS Security Hub")
```
Comment (lizrice) [inline cmd/root.go:160]:
```suggestion
	RootCmd.PersistentFlags().BoolVar(&aSFF, "asff", false, "Send the results to AWS Security Hub")
```

File: job-eks.yaml [hunk 14]
```diff
@@ -11,7 +11,7 @@ spec:
         - name: kube-bench
           # Push the image to your ECR and then refer to it here
           image: <ID.dkr.ecr.region.amazonaws.com/aquasec/kube-bench:ref>
-          command: ["kube-bench", "node", "--benchmark", "eks-1.0"]
+          command: ["kube-bench", "node", "--benchmark", "eks-1.0", "--assf"]
```
Comment (lizrice) [inline job-eks.yaml:14]:
```suggestion
          command: ["kube-bench", "node", "--benchmark", "eks-1.0", "--asff"]
```

**Review by lizrice — CHANGES_REQUESTED** [review]
Thank you! Most of my sugggestions here are to keep the name consistent with Amazon Security Finding Format (ASFF) but generally this is looking good!

File: cfg/eks-1.0/config.yaml [hunk 15]
```diff
@@ -1,2 +1,8 @@
 ---
 ## Version-specific settings that override the values in cfg/config.yaml
+## AWS account number is required.
+AWS_ACCOUNT: "<AWS_ACCT_NUMBER>"
+## AWS region is required.
+AWS_REGION: "<AWS_REGION>"
+## EKS Cluster ARN is required.
+CLUSTER_ARN: "<AWS_CLUSTER_ARN>"
```
[the remark this answers is not in the document]
Comment (brianterry) [reply cfg/eks-1.0/config.yaml:8]: Yes, you are correct.

File: cmd/common.go [hunk 16]
```diff
@@ -404,6 +408,18 @@ func writePgsqlOutput(controlsCollection []*check.Controls) {
 	}
 }
 
+func writeAASFOutput(controlsCollection []*check.Controls) {
+	for _, controls := range controlsCollection {
+		out, err := controls.AASF()
+		if err != nil {
+			exitWithError(fmt.Errorf("failed to output to AASF: %v", err))
```
[the remark this answers is not in the document]
Comment (brianterry) [reply cmd/common.go:415]: I agree

File: check/controls.go (hunk 2 above)
Comment (brianterry) [reply check/controls.go:203]: Yes. Agree that we should update to ASFF

File: cmd/securityHub.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+		return err
+	}
+	svc := securityhub.New(sess)
+	p := findings.New(svc)
+	out, perr := p.PublishFinding(in)
+	print(out)
+	if perr != nil {
+		return perr
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (brianterry) [reply cmd/securityHub.go:35]: Nice!

**Review by lizrice — APPROVED** [review]
Good to go I think!

**Comment by lizrice** [comment]
This is setting a terrible precedent but I am overriding the fact that the Travis builds haven't completed given that they have been pending for hours! Have confirmed that all integration and unit tests are running fine locally

**Pull request merged**